fixes #1923 - matches vms on a compute resource to hosts

jmontleon opened the pull request:
This lets you match vms to host records via mac address, either all at once by clicking on Match VM's to Hosts at the Compute Resource Level, or the Manage button, if available (there is logic to determine if it is capable of matching a host and will only show in the case that it can), at the compute resource vm.

jmontleon committed:
fixes #1923 - matches vms on a compute resource to hosts

ohadlevy commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,19 @@ def create
 
+  def manage
+    vms = @compute_resource.vms
+    Host.where(:uuid => NIL).each do |host|
```
why NIL in cap? ( i know its the same, but nil would work too :))

jmontleon replied:
If I had to guess it was because I was staring at NULL in the database all over the place when I was removing the compute_resource_id's and uuid's and looking at the table contents before writing the patch to be sure I understood what needed to be done. I think I just jumped from NULL to NIL.

domcleal commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,19 @@ def create
+  def manage
+    vms = @compute_resource.vms
+    Host.where(:uuid => nil).each do |host|
+      vm = vms.detect {|vm| vm.mac == host.mac if vm.respond_to?(:mac)}
```
I wonder if this is robust enough for multi-NIC VMs and hosts?  Maybe match any MAC on the VM against any on the host?

domcleal commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,19 @@ def create
+      if vm.present?
+        host.compute_resource_id = @compute_resource.id
+        host.uuid = vm.identity
+        host.save(:validate => false)
```
why disable validations here?

domcleal commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,19 @@ def create
+        host.save(:validate => false)
+      end
+    end
+    redirect_to compute_resource_path(:compute_resource_id => params[:compute_resource_id])
```
Maybe use "notice" to tell the user that X hosts have been associated?

domcleal commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -31,6 +31,14 @@ def show
   end
 
+  def manage
+    host = Host.where(:mac => @vm.mac).first
```
This might return no host?

domcleal replied:
Ah, I see you're restricting it in the view... ok

domcleal commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -31,6 +31,14 @@ def show
+    host.uuid = @vm.identity
+    host.compute_resource_id = @compute_resource.id
+    host.save(:validate => false)
+    redirect_to compute_resource_vm_path(:compute_resource_id => params[:compute_resource_id], :id => @vm.identity)
```
A notice would be good here too.  I'm wondering if it should redirect to the newly associated host, or add a link to it from the page at least.

domcleal commented on app/views/compute_resources/show.html.erb:
```diff
@@ -1,6 +1,7 @@
 <%= javascript "compute_resource" %>
 <% title @compute_resource.name %>
 
+<% title_actions display_link_if_authorized(_("Match VM's to Hosts"), hash_for_manage_compute_resource_path(:compute_resource_id => @compute_resource), :method => :put, :class=>"btn btn-success") %>
```
nitpick, no apostrophe and lower case "Hosts"

domcleal commented on app/views/compute_resources_vms/show.html.erb:
```diff
@@ -1,6 +1,9 @@
   <%= render "compute_resources_vms/show/#{@compute_resource.provider.downcase}" %>
 </div>
+<%  if Host.where(:uuid => @vm.identity).length == 0 && Host.where(:mac => @vm.mac).length == 1 %>
+  <% title_actions display_link_if_authorized(_("Manage"), hash_for_manage_compute_resource_vm_path(:compute_resource_id => @compute_resource, :id => @vm.id), :method => :put, :class=>"btn btn-success") %>
```
Ideally I think the text should be similar on the two pages so people realise they do the same thing.  Perhaps use the term "Associate", or "Manage" on both?

domcleal commented on app/views/compute_resources_vms/show.html.erb:
```diff
@@ -1,6 +1,9 @@
 <div class='row'>
   <%= render "compute_resources_vms/show/#{@compute_resource.provider.downcase}" %>
 </div>
+<%  if Host.where(:uuid => @vm.identity).length == 0 && Host.where(:mac => @vm.mac).length == 1 %>
```
ditto above, if the MAC->NIC matching does change to be more than just a single MAC, then this would need updating too.  For now I think it's OK.

domcleal commented:
There are two test failures as you've added routes but not linked them to permissions so non-admin users can make use of them.

See app/services/foreman/access_permissions.rb and add both "compute_resources/manage"
 and "compute_resources_vms/manage" to one of the permissions.. probably the edit_hosts permission?  You'll also need to use `display_link_if_authorized` or `authorized_via_my_scope` in the UI to prevent the button from appearing.

Here are the test results: http://ci.theforeman.org/job/test_develop_pull_request/3082/

domcleal commented:
For my own benefit, this relates to http://projects.theforeman.org/issues/1923

Could you update the commit message too to read "fixes #1923 - ....stuff" please?

Looks good though @jmontleon, thanks for tidying it up and sending it in.  You'll make a lot of people happy!

jmontleon commented:
@domcleal I think I've dealt with all the items you raised except the possibility of vms with multiple nics. I'm going to have to think and dive a little deeper on that one unless anyone has an idea.

domcleal commented:
That's looking great, thanks for the fast update @jmontleon.  I'm not wedded to the idea of multi-NIC support, but it seems like a likely issue we could encounter.

jmontleon commented:
@domcleal This needs to be changed some. It causes an error when trying to view an AWS compute resource because there is no @vm.mac. 

Also, if testing with AWS this will probably be the first thing you hit when you try to pull up the compute_resource_vm view for ec2.

I think Ohad said it the association worked for vmware, ovirt, and libvirt (these have a mac to use so it should, though I've only tested with ovirt myself)

If the ip address is a sufficient identifier for ec2 we can use :ip => @vm.public_ip_address for ec2.

Unsure about rackspace (think public_ip_address as well?) and openstack (floating_ip_address?) off the top of my head.

I've pushed another commit that should make the compute_resource_vms view work with vmware/ovirt/libvirt/ec2 . the compute_resource view is still broken at the moment, but should be easily changed in the same fashion, so I'll work on that tomorrow. 

I've also considered that it might be better to break it out in each CR types view but haven't done that so far.

edit: it should work from compute_resource or compute_resource_vms now for the 4 plus probably rackspace. Still unsure how to tackle openstack.

jmontleon commented:
split out the cr specific stuff to the appropriate views. added a little more info to the notices to give a hint how it's matching for each type of CR. 

Added the the means to match based on floating_ip_address for openstack; it seems at present foreman requires one to create an openstack host, however im not sure this is the best way possible since openstack itself doesn't require a floating ip, but is there some other common uniquely identifying information we can use?

domcleal commented:
I see what you're trying to do by handling each CR in its own view, but I think reimplementing the checks here is bad.  It would be better to skip the check to see if a possible host is known and always display the button if there's no host currently associated.  This way it could flag a message up to the user (from the controller) when they click "Associate" telling them on what criteria it failed to associate.  At the moment, the reason is hidden because the button simply isn't visible (and can cause CR-specific errors, as you found).

The association code should move to each CR model, so the controller code is smaller and more generic.  This will allow us to hook it up to the API easier in the future too.

@GregSutcliffe - any thoughts on the OpenStack stuff?  I'm not familiar with it.

I hope that's clear, ping me if not.

GregSutcliffe commented:
@domcleal I agree with your points re: the logic.

Specifically on OpenStack, @jmontleon that will work for now, but there's an open issue to remove the floating IP requirement in Foreman (assuming Foreman can reach the internal network). Presumably, in doing so, we'll be able to record the internal IP in Foreman (as EC2 does now, displaying both IPs on the Host) page, so it might make things actually easier.

jmontleon commented:
@domcleal not sure if this is what you had in mind. This makes the association happen in the model, but feels like there is a lot of repetitive code. I guess it could just find the host and return it for the controller to do the work, but I don't known what's better when it comes to hooking it into the API and keeping the controller code minimal, as you mentioned. 

And ignore my question on irc - since the button is back up on show.html.erb rather than the cr specific view it no longer shows on the host page. Also the button is no longer green.

domcleal commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,19 @@ def create
+  def associate
+    count = 0
+    @compute_resource.vms.each do |vm|
+      if Host.where(:uuid => vm.identity).length == 0 
```
nitpick, `.length == 0` can be written as `.empty?`

domcleal commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,19 @@ def create
+        end
+      end
+    end
+    notice _("#{count} VMs associated to hosts")
```
There's a trick to pluralisation of strings so they can be translated (see also [translating](http://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers)).  You'd write this as:

```
n_("%s VM associated to a host", "%s VMs associated to hosts", count) % count
```

(singular, plural, count and then the % replaces the %s with count in the resulting string)

domcleal commented:
@jmontleon sorry, I should have been clearer.  Your instinct is right, it's a lot of repetitive code.  Keep the host update and saving in the controller but on the CR models have a method (`associated_host`?) to return the host back to the controller.

jmontleon commented:
@domcleal another attempt :)

domcleal commented on app/models/compute_resources/foreman/model/ec2.rb:
```diff
@@ -91,6 +91,13 @@ def update_required?(old_attrs, new_attrs)
     end
 
+    def associated_host(uuid)
+      vm = find_vm_by_uuid(uuid)
```
Could you pass in `vm` itself rather than `vm.identity` to save a lookup?

domcleal commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,23 @@ def create
+    count = 0
+    @compute_resource.vms.each do |vm|
+      if Host.where(:uuid => vm.identity).empty?
+        host = @compute_resource.associated_host(vm.identity) if @compute_resource.respond_to?(:associated_host)
```
As this is called only for one CR, maybe move this check out of the loop and fail early if it doesn't respond to associated_host.

domcleal commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -24,6 +24,25 @@ def create
+        redirect_to host_path(host)
+      else
+        notice _("No host found to associate this vm with")
+        redirect_to compute_resource_vm_path(:compute_resource_id => params[:compute_resource_id], :id => @vm.identity)
```
Style note, look at process_success and process_error (particularly in the hosts_controller for examples) as it can combine the printing of a notice and a redirect just back to the last page (which is mostly what you're doing) or to a given path.

domcleal commented:
Great work @jmontleon, that's looking about right now.  My next step will be to test it and give you some feedback from that.

I'd also like to see a couple of tests added for the two associate methods as they're simpler now - we'd need them in:

a) `test/functional/compute_resources_controller_test.rb`, which would use an existing host (from `test/fixtures/hosts.yml`), then maybe stub `associated_host` to return a certain UUID and then check the host is updated with the CR ID and UUID.
b) `test/functional/compute_resources_controller_vms_test.rb`, for essentially the same

If it helps, I can try and get you an example to work from.

jmontleon commented:
@domcleal, regarding tests is something like this at least on the right track? I also don't know how to 'run' the test, so I guess I will figure that out tomorrow too and it will probably help to determine if it's anywhere near correct or not.

  test "should not associate host when not permitted" do
    setup_user "associate"
    put :associate, {:compute_resource_id => @your_compute_resource.to_param, :compute_resource => {:name => "associating_vm", :provider => "EC2"}, :vm => {:identity => "ami-abcdef01", :private_ip_address => "10.0.0.1", :public_ip_address => "2.3.4.5"}, set_session_user
    assert_response 403
  end

  test "should associate host" do
    setup_user "associate"
    put :associate, {:compute_resource_id => @your_compute_resource.to_param, :compute_resource => {:name => "associating_vm", :provider => "EC2"}, :vm => {:identity => "ami-abcdef01", :private_ip_address => "10.0.0.1", :public_ip_address => "2.3.4.5"}, set_session_user
    assert_redirected_to compute_resources_path
  end

domcleal commented:
Yes, that's the start of it.  Some of the parameters there (such as :vm) you don't need, but then in the second test what you need to do is to test the host object is updated, e.g.

```
assert_equal "...", host(:one).uuid
assert_equal compute_resource(:one).id, host(:one).compute_resource_id
```

You can run the full test suite with `rake test`, or individual test suites with `ruby -Itest test/functional/compute_resources_controller.rb`.

jmontleon commented:
@domcleal I think I've gotten a little over my head on this. What I've figured out is that the test vm in libvirt-1.0.5.4/src/test/test_driver.c does not have any interface definitions, and so no mac address, which is making it very difficult to associate it to a vm. 

Adding the xml for an interface seems to break the libvirt build tests at the end of compiling, so patching that doesn't seem like the greatest option either as it will likely mean fixing 11 tests on libvirt and working on getting that submitted upstream.

It seems all the test compute_resources are libvirt providers at the moment as well, so having spent most of the day trying to figure this bit out I haven't looked to see if one of the other models would conjure up something workable.

Is there a less crazy way to go about this? I feel like I'm doing it wrong :)

domcleal commented on app/models/compute_resources/foreman/model/ec2.rb:
```diff
@@ -91,6 +91,12 @@ def update_required?(old_attrs, new_attrs)
     end
 
+    def associated_host(vm)
+      host = Host.where(:ip => vm.public_ip_address).first
```
These should use Host.my_hosts to prevent it updating a host outside of the user's control.

domcleal commented on app/models/compute_resources/foreman/model/ec2.rb:
```diff
@@ -91,6 +91,12 @@ def update_required?(old_attrs, new_attrs)
 
+    def associated_host(vm)
+      host = Host.where(:ip => vm.public_ip_address).first
+      host = Host.where(:ip => vm.private_ip_address).first unless host.present?
```
Style nitpick: `host ||= Host.where` will only set "host" if it's false, so you don't need the unless clause.

domcleal commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -24,6 +24,22 @@ def create
+      host = @compute_resource.associated_host(@vm) if @compute_resource.respond_to?(:associated_host)
+      if host.present?
+        host.uuid = @vm.identity
+        host.compute_resource_id = @compute_resource.id
```
Nitpick, you can do `.compute_resource = @compute_resource`

isratrade replied:
@domcleal, I prefer the style that @jmontleon did for setting an *_id = @object.id rather than setting the object.  Although Rails is supposed to be smart enough to set the id, I've had some weird bugs in the past, so I usually just set the id explicitly.

jmontleon replied:
I changed it to host.compute_resource_id = @compute_resource and it started setting all compute_resource_id's to 1. I reverted it, but if I can get to the bottom of why that happened I can change it again.

domcleal commented:
Sorry for the delay and neglecting you!

That really is deep, I see what you mean.  I hadn't realised we were using libvirt's internal test infrastructure - I didn't know it existed.  There's a mocking framework from Fog that I thought we used, activate it by calling Fog.mock! in a test setup.

Here's a test using Fog mocking that seems to pass:
https://gist.github.com/domcleal/6320460

Do you think there needs to be something to prevent us changing hosts that are already associated?  It could help prevent weird accidents, but the capability might be useful to fix broken data in Foreman, say when a UUID somehow changes on a hypervisor.  I'm now actually tending towards not needing this, so perhaps ignore me.

domcleal commented on app/controllers/compute_resources_vms_controller.rb:
```diff
@@ -24,6 +24,22 @@ def create
+      if host.present?
+        host.uuid = @vm.identity
+        host.compute_resource_id = @compute_resource.id
+        host.save(:validate => false) # don't want to trigger callbacks
```
Use `.save!` here, it'll throw errors then (this caught me out).

domcleal commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,24 @@ def create
+          if host.present?
+            host.uuid = vm.identity
+            host.compute_resource_id = @compute_resource.id
+            host.save(:validate => false) # don't want to trigger callbacks
```
`.save!`

ohadlevy commented on app/controllers/compute_resources_controller.rb:
```diff
@@ -49,6 +49,24 @@ def create
+  def associate
+    count = 0
+    if @compute_resource.respond_to?(:associated_host)
+      @compute_resource.vms.each do |vm|
```
should this move into the model?

ohadlevy commented on app/models/compute_resources/foreman/model/ec2.rb:
```diff
@@ -91,6 +91,11 @@ def update_required?(old_attrs, new_attrs)
     end
 
+    def associated_host(vm)
+      host = Host.my_hosts.where(:ip => vm.public_ip_address).first
```
you can also do where(:ip => [vm.public_ip_address, vm.private_ip_address])

assuming there is only one host with those ips, it should work.

ohadlevy commented on app/models/compute_resources/foreman/model/openstack.rb:
```diff
@@ -70,6 +70,11 @@ def console(uuid)
     end
 
+    def associated_host(vm)
+      host = Host.my_hosts.where(:ip => vm.floating_ip_address).first
```
same here

ohadlevy commented on app/models/compute_resources/foreman/model/rackspace.rb:
```diff
@@ -91,6 +91,11 @@ def provider_friendly_name
 
+    def associated_host(vm)
+      host = Host.my_hosts.where(:ip => vm.public_ip_address).first
+      host ||= Host.my_hosts.where(:ip => vm.private_ip_address).first
```
and here

domcleal commented:
Merged as 805358d for Foreman 1.3.  Thanks @jmontleon, this rocks.

lzap commented:
Release notes draft:

#### Bare metal migration

VM and host records can be now matched via MAC address, either all at once by
clicking on Match VM's to Hosts at the Compute Resource Level, or the Manage
button, if available at the compute resource VM. There is logic to determine
if it is capable of matching a host.
